Source/CarthageKit/Xcode.swift

```diff
@@ -971,31 +984,52 @@ public func buildInDirectory(directoryURL: NSURL, withConfiguration configuratio
 				return false
 			}
 		}
-		|> take(1)
-		|> flatMap(.Merge) { (project: ProjectLocator) -> SignalProducer<String, CarthageError> in
+		|> flatMap(.Concat) { (project: ProjectLocator) -> SignalProducer<[SharedScheme], CarthageError> in
 			return schemesInProject(project)
+				|> flatMap(.Merge) { scheme -> SignalProducer<SharedScheme, CarthageError> in
+					let buildArguments = BuildArguments(project: project, scheme: scheme, configuration: configuration)
+
+					return shouldBuildScheme(buildArguments, platform)
+						|> filter { $0 }
+						|> map { _ in SharedScheme(schemeName: scheme, project: project) }
+				}
+				|> collect
+				|> flatMap(.Concat) { schemes in
+					if schemes.isEmpty {
+						return .empty
+					} else {
+						return SignalProducer(value: schemes)
+					}
+				}
 		}
+		|> take(1)
+		|> flatMap(.Merge) { schemes in SignalProducer(values: schemes) }
 		|> combineLatestWith(locatorProducer |> take(1))
-		|> map { (scheme: String, project: ProjectLocator) -> BuildSchemeProducer in
-			let buildArguments = BuildArguments(project: project, scheme: scheme, configuration: configuration)
-
-			return shouldBuildScheme(buildArguments, platform)
-				|> filter { $0 }
-				|> flatMap(.Merge) { _ in
-					let initialValue = (project, scheme)
-
-					let buildProgress = buildScheme(scheme, withConfiguration: configuration, inProject: project, workingDirectoryURL: directoryURL)
-						// Discard any existing Success values, since we want to
-						// use our initial value instead of waiting for
-						// completion.
-						|> map { taskEvent in
-							return taskEvent.map { _ in initialValue }
-						}
-						|> filter { taskEvent in taskEvent.value == nil }
+		|> map { (scheme: SharedScheme, project: ProjectLocator) -> BuildSchemeProducer in
+			/// We should consider a case that the project locator passed in is
+			/// not a workspace, nor the project in which the scheme is defined.
+			let projectToUse: ProjectLocator
+			switch project {
+			case .Workspace:
+				projectToUse = project
 
-					return BuildSchemeProducer(value: .Success(Box(initialValue)))
-						|> concat(buildProgress)
+			case .ProjectFile:
+				projectToUse = scheme.project
+			}
```

Comment:
This doesn't seem quite right to me. It seems like we should either (a) choose the first workspace or (b) choose the first workspace that includes that scheme, failing back in either case to building from the project where the scheme was defined.

Also… I know you didn't write it, but I think it'd be clearer replace the `combineLatestWith` with a `flatMap` and move the this logic into that block.